enhancements/network/on-prem-service-load-balancers.md

```diff
@@ -229,35 +230,101 @@ doing all of your testing in a virtual network built on top of it.
-(CCM). An example of this is the [packet.net
-CCM](https://github.com/packethost/packet-ccm), which ensures MetalLB is
-deployed and also configures it properly to work in packet.net’s BGP
-environment.
-
-These integration options must be explored in more detail as part of a more
-detailed integration proposal.
+#### Configuration Interfaces
+
+MetalLB's current configuration interface is a single `ConfigMap`.  We only
+want to support CRD based configuration through OpenShift.  The ultimate goal
+will be to enhance MetalLB with a native CRD based configuration interface.
+The design work has begun here:
+
+https://github.com/metallb/metallb/pull/833
+
+Since this is an upstream effort, the completion of this work may take longer
+than the timeline for when we want to ship the first piece of MetalLB support
+for OpenShift, which will be to support MetalLB in layer2 mode only.
+
+To mitigate this timing risk, we have another option for a way to configure the
+layer2 mode of MetalLB using an existing OpenShift configuration option.  As a
+backup plan, we can have the OpenShift version of the MetalLB operator build on
+top of the existing feature to automatically assign external IPs to services of
+`type=LoadBalancer`.  For example, take this sample configuration:
+
+```yaml
+apiVersion: config.openshift.io/v1
+kind: Network
+metadata:
+  name: cluster
+spec:
+  ...
+  externalIP:
+    autoAssignCIDRs:
+    - 10.10.10.0/24
+```
+
+This makes OpenShift automatically set the `externalIP` field, but nothing will
+make sure those addresses are reachable automatically.  That's where MetalLB
+comes in.  If the OpenShift MetalLB operator is enabled, the operator can
+automatically generate a MetalLB config map based on this cluster
+configuration.  The operator will read the cluster network configuration and
+generate config like this:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  namespace: metallb-system
+  name: config
+data:
+  config: |
+    address-pools:
+    - name: openshift
+      protocol: layer2
+      auto-assign: false
+      addresses:
+      - 10.10.10.0/24
+```
+
+Note how the configuration copies over the CIDR for the externalIP allocator
+used in OpenShift.  This makes MetalLB expect to see addresses in this range
+used for load balancer Services.  Also note the use of `auto-assign: false`.
+In this scenario, we're letting OpenShift do the address allocation piece.
+
+If we support this as a configuration interface, customers will be able to opt
+for the more feature rich CRD interface offered by MetalLB natively
+once it is ready.  As long as the `AutoAssignCIDRs` feature is not enabled on
+the OpenShift side, there should not be a conflict.
+
+**TODO** -- There is one outstanding problem to make this fallback plan work.
+The OpenShift allocator sets spec.ExternalIPs, while MetalLB expects that an
+external IP allocator will set spec.LoadBalancerIP.  We must fix this mismatch
+on one side or the other.
```

Comment:
ok this specific to auto-assign false, in which k8s resource is that , I didn't have to mess with that so I have no idea where that need to be set ?

Reply:
See this code in the metallb controller: https://github.com/metallb/metallb/blob/main/controller/service.go#L89-L96

It supports something else performing the IPAM function, but the field it looks at to see what IP has been requested is `LoadBalancerIP`, and not `ExternalIPs`, which is what the OpenShift IPAM knows how to set.  We need to reconcile this somehow.

Choices include ...

1) Change MetalLB to also look at spec.ExternalIPs (but this doesn't seem correct to me)

2) Change OpenShift to optionally set LoadBalancerIP instead.  It could be by duplicating the `AutoAssignCIDRs` functionality with a near identical option called `AutoAssignLBCIDRs` or something like that.

3) Write a controller that copies the IP from `ExternalIPs` to `LoadBalancerIP` on Services when we know we're running in this mode.  This feels pretty ugly, but would probably work ...

Reply:
1. Agreed, I don't think this is correct either
2. With potentially different CIDRs?
3. A bit ugly but it was the first solution that I considered.

Reply:
2 this will require changes in the API repo  to add another field ?
AutoAssignLBCIDRs []string `json:"address-pools,omitempty"`

Reply:
On 2) 
* Maybe we could raise an alert if both options were configured.  I don't think it makes sense for both to be used at the same time.
* Yes, this would be a new API addition as you describe.

3 seems like the "easiest" solution.  We could do this entirely in the OpenShift metallb operator and it would be part of how it works when it's operating in this "auto-configure based on AutoAssignCIDRs" mode.  It just feels like an ugly hack.

I think my preference is #2.  It seems the most explicit and clean to me.  It seems like we should have time to get this in.

Reply:
I agree, Option 2, (with a suitable config flag) seems like the cleanest and easiest to implement.

Reply:
Thanks, all.  The PR has been updated to reflect option 2 as the proposed approach.

It includes a proposed option name of `AutoAssignLoadBalancerCIDRs`, but I'm not tied to the name at all.